From b0f64aae20c58c733e90c8b82228ecfcbc1f2f17 Mon Sep 17 00:00:00 2001
From: Scott Moser <smoser@ubuntu.com>
Date: Fri, 16 Jan 2015 14:29:48 -0500
Subject: [PATCH] hostname: apply hostname same as is written

on RHEL, we were writing to persistent configuration the fqdn, but
invoking 'hostname' on the first boot with just the shortname.  On 'reboot',
then the hostname would differ.

Now, whatever we write, invoke hostname with.

Also remove some duplicate code.
---
 cloudinit/distros/__init__.py | 9 ++++++---
 cloudinit/distros/arch.py     | 7 -------
 cloudinit/distros/debian.py   | 7 -------
 cloudinit/distros/freebsd.py  | 5 -----
 cloudinit/distros/gentoo.py   | 7 -------
 cloudinit/distros/sles.py     | 7 -------
 6 files changed, 6 insertions(+), 36 deletions(-)

diff --git a/cloudinit/distros/__init__.py b/cloudinit/distros/__init__.py
index 55d6bcb..74469bb 100644
--- a/cloudinit/distros/__init__.py
+++ b/cloudinit/distros/__init__.py
@@ -85,7 +85,7 @@ class Distro(object):
     def set_hostname(self, hostname, fqdn=None):
         writeable_hostname = self._select_hostname(hostname, fqdn)
         self._write_hostname(writeable_hostname, self.hostname_conf_fn)
-        self._apply_hostname(hostname)
+        self._apply_hostname(writeable_hostname)
 
     @abc.abstractmethod
     def package_command(self, cmd, args=None, pkgs=None):
@@ -159,9 +159,12 @@ class Distro(object):
             util.logexc(LOG, "Failed to non-persistently adjust the system "
                         "hostname to %s", hostname)
 
-    @abc.abstractmethod
     def _select_hostname(self, hostname, fqdn):
-        raise NotImplementedError()
+        # Prefer the short hostname over the long
+        # fully qualified domain name
+        if not hostname:
+            return fqdn
+        return hostname
 
     @staticmethod
     def expand_osfamily(family_list):
diff --git a/cloudinit/distros/arch.py b/cloudinit/distros/arch.py
index 310c3df..3252af8 100644
--- a/cloudinit/distros/arch.py
+++ b/cloudinit/distros/arch.py
@@ -120,13 +120,6 @@ class Distro(distros.Distro):
                 return False
         return True
 
-    def _select_hostname(self, hostname, fqdn):
-        # Prefer the short hostname over the long
-        # fully qualified domain name
-        if not hostname:
-            return fqdn
-        return hostname
-
     def _write_hostname(self, your_hostname, out_fn):
         conf = None
         try:
diff --git a/cloudinit/distros/debian.py b/cloudinit/distros/debian.py
index 1ae232f..ca8d208 100644
--- a/cloudinit/distros/debian.py
+++ b/cloudinit/distros/debian.py
@@ -88,13 +88,6 @@ class Distro(distros.Distro):
         else:
             return distros.Distro._bring_up_interfaces(self, device_names)
 
-    def _select_hostname(self, hostname, fqdn):
-        # Prefer the short hostname over the long
-        # fully qualified domain name
-        if not hostname:
-            return fqdn
-        return hostname
-
     def _write_hostname(self, your_hostname, out_fn):
         conf = None
         try:
diff --git a/cloudinit/distros/freebsd.py b/cloudinit/distros/freebsd.py
index d98f957..d15930c 100644
--- a/cloudinit/distros/freebsd.py
+++ b/cloudinit/distros/freebsd.py
@@ -92,11 +92,6 @@ class Distro(distros.Distro):
             return default
         return hostname
 
-    def _select_hostname(self, hostname, fqdn):
-        if not hostname:
-            return fqdn
-        return hostname
-
     def _write_hostname(self, hostname, filename):
         self.updatercconf('hostname', hostname)
 
diff --git a/cloudinit/distros/gentoo.py b/cloudinit/distros/gentoo.py
index 09f8d8e..0fd8f06 100644
--- a/cloudinit/distros/gentoo.py
+++ b/cloudinit/distros/gentoo.py
@@ -99,13 +99,6 @@ class Distro(distros.Distro):
         else:
             return distros.Distro._bring_up_interfaces(self, device_names)
 
-    def _select_hostname(self, hostname, fqdn):
-        # Prefer the short hostname over the long
-        # fully qualified domain name
-        if not hostname:
-            return fqdn
-        return hostname
-
     def _write_hostname(self, your_hostname, out_fn):
         conf = None
         try:
diff --git a/cloudinit/distros/sles.py b/cloudinit/distros/sles.py
index 9788a1b..43682a1 100644
--- a/cloudinit/distros/sles.py
+++ b/cloudinit/distros/sles.py
@@ -115,13 +115,6 @@ class Distro(distros.Distro):
         conf.set_hostname(hostname)
         util.write_file(out_fn, str(conf), 0644)
 
-    def _select_hostname(self, hostname, fqdn):
-        # Prefer the short hostname over the long
-        # fully qualified domain name
-        if not hostname:
-            return fqdn
-        return hostname
-
     def _read_system_hostname(self):
         host_fn = self.hostname_conf_fn
         return (host_fn, self._read_hostname(host_fn))
